### ISSUE TITLE
longrunning: find an import path outside of cloud

### ISSUE THREAD

**Issue by broady** [GoogleCloudPlatform/google-cloud-go#424]
The longrunning package is heavily tied to the standard Google lro proto, which makes it impossible to use with legacy LROs like BigQuery's Job.

I have a few proposed changes:
* Change Operation to an interface.
* Change Poll, Wait and Metadata to take an `interface{}` rather than a `proto.Messsage`
* Move the InternalNewOperation function to internal.

Happy to make these changes if you agree.

**Comment by zombiezen** [comment]
Are there examples other than BigQuery?

**Comment by rakyll** [comment]
> Are there examples other than BigQuery?

**Comment by rakyll** [comment]
Speech package will also require long running jobs.

Also cc'ing @pongad. (We had a conversation about async response handling APIs with long running jobs before.)

**Comment by pongad** [comment]
@rakyll Thank you for looping me in. I think the [speech](https://godoc.org/cloud.google.com/go/speech/apiv1beta1) API is already using LRO proto, so I don't think it will further benefit from the proposed change. Are we looking at different versions of the API?

Having said that, since this helps BigQuery and possibly others, I don't see a reason to not do it.

I think the only package that depends on longrunning is speech/apiv1beta1. It is automatically generated by toolkit. To keep the build working, we might want to hand edit the generated code (`despiteallobjections` written at the top of the file). I can make a separate change to toolkit for future clients.

**Comment by broady** [comment]
I think BigQuery is the major one, but this change also lets us hide the `InternalNewOperation` function, which is distracting.

Changing it to an interface should be straightforward - no struct fields are exported.

**Comment by jba** [comment]
I'm not opposed to this change. Go ahead and make a CL.

However, I don't see the advantage of bigquery.Job implementing longrunning.Operation:
* Job has an ID method, and we have a corresponding Client.JobFromID. Now Job would also (or instead?) have a Name method, and I guess we'd have Client.JobFromName. I'm pretty sure they're different things: Name is the full resource name ("projects/P/jobs/J") whereas ID is just the last component of the name. So all in all, this would be confusing.

* Job has a Status method that returns a JobStatus. It would now also (or instead?) have a Metadata method, which takes an interface{} into which you must pass in a JobStatus pointer. A clumsier and more obscure surface.

* Job has no Poll or Wait methods (on my to-do list). If it did, though, they'd return *JobStatus, rather than take a *JobStatus as an interface{} argument. Same issue as the Job.Status method.

* If we did all this, what would we gain? What could we do with a Job in virtue of its implementing longrunning.Operation that we couldn't do before?

Bottom line: longrunning.Operation is a slightly ugly package that has to be that way to work within the general long-running operation framework. For any particular API that already has its own long-running op equivalent, we should use similar method names and semantics (I will base Job.Poll/Wait's on longrunning's) without trying to match the exact interface.

**Comment by rakyll** [comment]
> we should use similar method names and semantics (I will base Job.Poll/Wait's on longrunning's) without trying to match the exact interface.

+1 to this idea. Naming conventions are often useful if we cannot match the exact API.

@pongad, any ideas whether longrunning should be a top-level package under the Cloud packages? It is more of a support package and works at the level of proto.Message. Maybe we need to generate a more user-friendly surface layer that works with concrete types and make the longrunning package entirely an internal support package.

**Comment by pongad** [comment]
>Maybe we need to generate a more user-friendly surface layer that works with concrete types and make the longrunning package entirely an internal support package.

@rakyll I struggled with this question myself while writing the package. Generating surface layer with concrete types is certainly helpful. On the other hand, I also think it's important to avoid spamming the user with too many types and methods.

A reasonable middle ground might look something like this:
- Move `cloud.google.com/go/longrunning` into `cloud.google.com/go/support/longrunning`.
Not exactly internal but a little more out of the way.

- Rename `Operation.Wait` to `Operation.WaitProto`.

- For methods that return `Operation` that will be completed with `MyMessage`, generate something like 
```go
type MyMessageOperation struct {
  *longrunning.Operation
}

func (o *MyMessageOperation) Wait(ctx context.Context) (*MyMessage, error) {
  msg := new(MyMessage)
  if err := o.WaitProto(ctx, msg); err != nil {
    return nil, err
  }
  return msg, nil
}
```

I'm working under the assumption that most users will just want to wait for the response and not care too much about metadata. I think this the assumption feels reasonable, but I don't have hard data to prove it.

In this way, the common case has a more friendly surface with concrete types. More advanced use cases are still supported through the `Operation` member without having to generate a swath of code each time.

**Comment by jba** [comment]
Thought about this some more. OK with moving it to a support subdir, but I don't yet see the value of the other changes. The wrappers that pongad is generating deal with most cases. I haven't yet seen an example where making Operation an interface or changing the arg types from proto.Message to interface{} would add value.

**Comment by adams-sarah** [comment]
Is this issue still relevant?

**Comment by rakyll** [comment]
@pongad, the longrunning wrappers sound pretty good to me.

If we will expose the wrappers, maybe we really can move the longrunning into internal or document that users often don't need to use that package directly.

> The wrappers that pongad is generating deal with most cases. I haven't yet seen an example where making Operation an interface or changing the arg types from proto.Message to interface{} would add value.

If we are generating nice wrappers around it, I agree that interface{} would not make any sense.

> Is this issue still relevant?

I think we are still using this issue as a tracker related to the longrunning jobs. The wrappers are still being debated AFAIK.

**Comment by pongad** [comment]
Sorry I lost track of this. The concrete-typed wrappers have been implemented for a while now. However, when we generate non-cloud APIs, some APIs might need to use this package. It would be a little weird for non-cloud packages to depend on cloud repo, but at least things will keep working until we find longrunning a new home. I think we should punt on this until we're more sure where to move the package to.

**Comment by jba** [comment]
I've been working on finding an import path prefix for non-cloud clients. I'll push on that a little harder.

At that point the best we can probably do for a while is copy the package. We'll need everyone (in particular AppEngine) to have type aliases before we can dedup.

**Comment by jba** [comment]
Moved to https://github.com/GoogleCloudPlatform/google-cloud-go/wiki/Long-Term-Plans.
